storage/reader.go

```diff
@@ -392,6 +574,72 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return n, err
 }
 
+func (r *Reader) readWithGRPC(p []byte) (int, error) {
+	// No stream to read from, either never initiliazed or Close was called.
+	// TODO: should an error be returned here? "reader has been closed"?
+	if r.stream == nil {
+		return 0, io.EOF
+	}
+
+	// The entire object has been read by this reader, return EOF.
+	if r.size != 0 && r.size == r.seen {
+		return 0, io.EOF
+	}
+
+	n := copy(p, r.leftovers)
+	r.seen += int64(n)
+	r.leftovers = r.leftovers[n:]
+
+	for len(p[n:]) > 0 {
+		msg, err := r.stream.Recv()
+		if err != nil {
+			if err == io.EOF && r.seen == r.size {
+				// Let the next call to readWithGRPC return EOF.
+				return n, nil
+			}
+			// This prevents reopening of the stream when the error is not
+			// retryable.
+			if !shouldRetry(err) {
+				return n, err
+			}
+
+			// Save the original error to return instead of the reinit error.
+			e := err
+
+			// Close existing stream and initialize new stream with updated
+			// offset.
+			r.closeStream()
+
+			// This will immediately attempt to reopen the stream, but will
+			// backoff if further attempts are necessary.
+			var res *readStreamResponse
+			res, r.cancelStream, err = r.reopenWithGRPC(r.seen)
+			if err != nil {
+				// Cannot reinstantiate the stream so return the original error.
+				return n, e
+			}
+			r.stream = res.stream
+			// Reopened the stream and Recv'd the first message, so carry on.
+			msg = res.response
+		}
+
+		// TODO: Capture incremental CRC32C for this chunk.
```

Comment:
Can we make a longer TODO about what we want to do for per-chunk checksums?

Reply:
Done! I don't really know how to do this or if it is necessary, because we already have full Object checksum verification in this PR, but I've expanded this comment nonetheless :)

Reply:
drive-by: the object checksums don't help for partial downloads (ranges, offsets, etc.) the per-chunk checksums would detect bit errors in that case.